### PR TITLE
[30412] Login.gov IAL2 implementation

### PR DESCRIPTION
**Pull request by bramleyjl**
## Description of change

This is the second part of our initial Login.gov rollout; our initial implementation will use the double-verification process currently in use with ID.me, except with IAL1/IAL2. For Login.gov ISAM is returning a `verified_at` attribute that indicates if the user was verified at IAL2 with Login.gov, the presence of that attribute on a currently-IAL1 account triggers the upcerting `verify` request to return the IAL2 user.

## Original issue(s)
department-of-veterans-affairs/va.gov-team#30412

## Things to know about this PR

SAML responses from ISAM are as follows:

IAL1:
```
  {"va_eauth_csid"=>["LOGINGOV"],
   "va_eauth_lastname"=>["NOT_FOUND"],
   "va_eauth_credentialassurancelevel"=>["1"],
   "va_eauth_ial"=>["1"],
   "va_eauth_firstname"=>["NOT_FOUND"],
   "va_eauth_csponly"=>["true"],
   "va_eauth_aal"=>["2"],
   "va_eauth_emailaddress"=>["test@test.com"],
   "va_eauth_transactionid"=>["nCebhF/mF31nS6tX8sEgYKymW8aASHsaxzTryAtFcPo="],
   "va_eauth_uid"=>["65f9f3b5-5449-47a6-b272-9d6019e7c2e3"],
   "va_eauth_issueinstant"=>["2021-11-10T22:23:11Z"],
   "va_eauth_middlename"=>["NOT_FOUND"],
   "va_eauth_verifiedAt"=>["2021-10-28T23:54:46Z"]}>
   ```
   IAL2:
   ```
     {"va_eauth_phone"=>["(999)999-9999"],
   "va_eauth_lastname"=>["USER"],
   "va_eauth_ial"=>["2"],
   "va_eauth_icn"=>["1200049153V217987"],
   "va_eauth_city"=>["Albany"],
   "va_eauth_country"=>["NOT_FOUND"],
   "va_eauth_csp_identifier"=>["200VLGN"],
   "va_eauth_gender"=>["NOT_FOUND"],
   "va_eauth_street2"=>["NOT_FOUND"],
   "va_eauth_aal"=>["2"],
   "va_eauth_csp_method"=>["LOGINGOV"],
   "va_eauth_dodedipnid"=>["NOT_FOUND"],
   "va_eauth_emailaddress"=>["john.bramley@adhocteam.us"],
   "va_eauth_cspid"=>["200VLGN_65f9f3b5-5449-47a6-b272-9d6019e7c2e3"],
   "va_eauth_issueinstant"=>["2021-11-10T22:26:49Z"],
   "va_eauth_birthDate_v1"=>["19870818"],
   "va_eauth_middlename"=>["TEST"],
   "va_eauth_state"=>["NY"],
   "va_eauth_birlsfilenumber"=>["NOT_FOUND"],
   "va_eauth_postalcode"=>["11111"],
   "va_eauth_street3"=>["NOT_FOUND"],
   "va_eauth_proofingAuthority"=>["FICAM"],
   "va_eauth_pid"=>["NOT_FOUND"],
   "va_eauth_csid"=>["LOGINGOV"],
   "va_eauth_credentialassurancelevel"=>["3"],
   "va_eauth_pnidtype"=>["SSN"],
   "va_eauth_mcid"=>["WSSOE2111101726506660012223612"],
   "va_eauth_firstname"=>["JOHN"],
   "va_eauth_prefix"=>["NOT_FOUND"],
   "va_eauth_street"=>["123 4th St"],
   "va_eauth_csponly"=>["false"],
   "va_eauth_pnid"=>["123123123"],
   "va_eauth_commonname"=>["john.bramley@adhocteam.us"],
   `"va_eauth_transactionid"=>["LcbO7Cubiu6cA2rKrTSL2UCstxkCqB10RuoKKwAyggI="],
   "va_eauth_suffix"=>["NOT_FOUND"],
   "va_eauth_uid"=>["65f9f3b5-5449-47a6-b272-9d6019e7c2e3"],
   "va_eauth_isDelegate"=>["false"],
   "va_eauth_secid"=>["1200049153"],
   "va_eauth_gcIds"=>["1200049153V217987^NI^200M^USVHA^P|65f9f3b5-5449-47a6-b272-9d6019e7c2e3^PN^200VLGN^USDVA^A|1200049153^PN^200PROV^USDVA^A|67f687a8ecd3448fbed4e5489b7eafc9^PN^200VIDM^USDVA^A"],
   "va_eauth_persontype"=>["NOT_FOUND"],
   "va_eauth_npi"=>["NOT_FOUND"],
   "va_eauth_street1"=>["123 4th St"],
   "va_eauth_mhvien"=>["65f9f3b5-5449-47a6-b272-9d6019e7c2e3"],
   "va_eauth_verifiedAt"=>["2021-10-28T23:54:46Z"]}>
   ```

Unit tests updated, manual login/logout with Login.gov at IAL1 & IAL2 & ID.me at LoA1 & LoA3 performed.

### CODE REVIEW

File: lib/saml/post_url_service.rb [hunk 1]
```diff
@@ -39,7 +39,7 @@ def initialize(saml_settings, session: nil, user: nil, params: {}, loa3_context:
     end
 
     def should_uplevel?
-      user.loa[:current] < user.loa[:highest]
+      user.loa[:current] < user.loa[:highest] || (!user.verified_at.nil? && user.loa[:current] < 3)
```
Comment (bosawt) [inline lib/saml/post_url_service.rb:42]: we should use the LOA::THREE constant here

File: lib/saml/url_service.rb [hunk 2]
```diff
@@ -138,8 +141,14 @@ def verify_url
           build_authn_context('myhealthevet_loa3')
         when 'dslogon', 'dslogon_multifactor'
           build_authn_context('dslogon_loa3')
+        when IAL::LOGIN_GOV_IAL1
+          build_authn_context([IAL::LOGIN_GOV_IAL2, AAL::LOGIN_GOV_AAL2], AuthnContext::LOGIN_GOV)
         when SAML::UserAttributes::SSOe::INBOUND_AUTHN_CONTEXT
```
Comment (bosawt) [inline lib/saml/url_service.rb:146]: This is fine for this PR, but I'm worried about using `SAML::UserAttributes::SSOe::INBOUND_AUTHN_CONTEXT` going forward. Can we create a ticket to remove that and refactor this line?

File: spec/lib/saml/ssoe_user_spec.rb [hunk 3]
```diff
@@ -93,10 +93,10 @@
 
       it 'has various important attributes' do
         expect(subject.to_hash).to eq(
-          email: 'vets.gov.user+262@example.com',
-          first_name: 'JERRY',
+          email: nil,
```
Comment (bosawt) [inline spec/lib/saml/ssoe_user_spec.rb:96]: Why are these nil? Is this expected?

**Review by bosawt — COMMENTED** [review]
Some quick comments, I'll test this right now but releasing the comments first

File: lib/saml/url_service.rb (hunk 2 above)
Comment (bramleyjl) [reply lib/saml/url_service.rb:146]: Right, plan on bringing it up at the meeting later today.

**Comment by bosawt** [comment]
Tested with ID.me sample user and Login.gov sample user, was able to log in with both (login.gov user required two user-input authentications but that is expected at this point and something we are working on outside of this PR)

File: spec/lib/saml/ssoe_user_spec.rb (hunk 3 above)
Comment (bramleyjl) [reply spec/lib/saml/ssoe_user_spec.rb:96]: Yes, we're not getting an email back for IAL1 users. I'll update the PR with the responses for IAL1 & 2.

Comment (bramleyjl) [reply spec/lib/saml/ssoe_user_spec.rb:96]: Turns out this is just an issue of adding `email` as a requested attribute, then it will return for IAL1

**Review by ericboehs — APPROVED** [review]

**Comment by nsoutouras** [comment]
MSI - Currently trying to get into master branch by EOD.  Should finish this sprint.

File: lib/saml/post_url_service.rb [hunk 4]
```diff
@@ -39,7 +39,7 @@ def initialize(saml_settings, session: nil, user: nil, params: {}, loa3_context:
     end
 
     def should_uplevel?
-      user.loa[:current] < user.loa[:highest]
+      user.loa[:current] < user.loa[:highest] || (!user.verified_at&.empty? && user.loa[:current] < LOA::THREE)
```
Comment (bosawt) [inline lib/saml/post_url_service.rb:42]: I wonder if we should refactor this just a bit to make it more clear what's happening here, maybe something like:
`user.verified_at.present? ? user.loa[:current] < LOA::THREE : user.loa[:current] < user.loa[:highest]`

File: lib/saml/url_service.rb [hunk 5]
```diff
@@ -138,8 +142,14 @@ def verify_url
...
-          "#{@user.identity.sign_in[:service_name]}_loa3"
+          if @type == 'logingov' || sign_in_service == 'logingov'
```
Comment (bosawt) [inline lib/saml/url_service.rb:148]: Were you going to move this into the `type` comparison from the other PR?

**Review by bosawt — COMMENTED** [review]
Just some small comments. I think you were going to roll the `url_service` `type` changes into this as well, yeah?

**Review by bosawt — APPROVED** [review]
Looks good! Tested with an id.me account and a login.gov account and both were able to log in and up level to LOA3

**Review by ericboehs — APPROVED** [review]

**Pull request merged**